### PR TITLE
#159204728 Completes pagination support on all articles, articles belonging to an author and single article endpoints

### PR DESCRIPTION
**Pull request by chukwuemekachm**
#### What does this PR do?
- Completes pagination support on all articles, articles belonging to an author and single article endpoints
#### Description of Task to be completed?
- Write unit tests for the endpoints
- Create the appropriate controllers and methods
- Use POSTMAN to test the endpoints
#### How should this be manually tested?
- Clone the repository
- Start up the command line /Terminal
- Change to the repo directory
- Run `npm install`
- Run `npm run migrate`
- Run `npm run dev`
- Use POSTMAN to create a GET request to `/api/articles?limit=2&offset=2`
- Use POSTMAN to create a GET request to `/api/users/:userId/articles?limit=2&offset=2`
- Use POSTMAN to create a GET request to `/api/articles/:slug`
#### Any background context you want to provide?
- Create a Postrgresql database before you start
#### What are the relevant pivotal tracker stories?
[#159204728](https://www.pivotaltracker.com/story/show/159204728)
#### Screenshots (if appropriate)

screen shot 2018-08-14 at 8 13 25 am
screen shot 2018-08-14 at 8 14 10 am

#### Questions:
- None

### CODE REVIEW

File: server/controllers/ArticleController.js [hunk 1]
```diff
@@ -13,4 +22,106 @@ export default class ArticleController {
...
+      if (article) {
+        return res.status(200).json({
+          status: 'success',
+          article,
+        });
+      }
+      return res.status(404).json({
+        status: 'fail',
+        errors: {
+          article: [`Article with slig: ${slug} not found.`],
```
Comment (yomigeek) [inline server/controllers/ArticleController.js]: Good work @chukwuemekachm, I think you needed to type slug not 'slig'.

**Review by Oluwafayokemi — APPROVED** [review]
Nice Job Chima! Well done.

File: server/controllers/ArticleController.js [hunk 2]
```diff
@@ -161,4 +167,111 @@ export default class ArticleController {
...
+  static sendPaginationResponse(res, articles, userId) {
+    // Returns a 20 when a list of articles are requested and found
+    if (articles[0]) {
+      return res.status(200).json({
+        status: 'success',
+        articles,
+      });
+    }
+    // Returns a 404 when a list of user articles are requested but not found
+    if (userId) {
```
Comment (Veraclins) [inline server/controllers/ArticleController.js]: Hi, @chukwuemekachm good job here. I particularly love how you simplified the whole process with helper functions. I am only consigned about this part of the code. The logic of returning an error if there is a userId is difficult to understand. I see that you are using it to differentiate the error messages for a single article and multiple articles but maybe you can consider using another method for the error situations which will use an argument that will not be misleading.

**Review by Veraclins — CHANGES_REQUESTED** [review]
@chukwuemekachm Good job Bro. Just look into the comment I made and you are good to go.

**Review by yomigeek — APPROVED** [review]
@chukwuemekachm I am satisfied with the job done, just ensure to implement the changes requested above.

File: server/controllers/ArticleController.js (hunk 2 above)
Comment (chukwuemekachm) [reply server/controllers/ArticleController.js]: Thanks, @Veraclins, I know the logic is a bit confusing. I need the argument to assert that the request was made for a list of articles belonging to an author. If the operation fails, I then use the `userId` argument to send a personalized message to the user. But I will try to refactor that part.
Thanks.

**Review by Veraclins — APPROVED** [review]

File: server/tests/singleArticle.test.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        res.body.article.should.have.property('categoryId');
+        res.body.article.should.have.property('body');
+        res.body.article.should.have.property('imageUrl');
+        res.body.article.should.have.property('createdAt');
+        res.body.article.should.have.property('updatedAt');
+        res.body.article.should.have.property('User');
+        res.body.article.should.have.property('ratings');
+        res.body.article.should.have.property('tags');
+        res.body.article.User.should.be.a('object');
+        res.body.article.tags.should.be.a('array');
```
Comment (madeofhuman) [inline server/tests/singleArticle.test.js:32]: Duuuuude!!! This is a very comprehensive test! 🙌🏾
However, did you need to test all the properties? Just a couple of unique properties would have achieved the same result. Nice work, though.

**Pull request merged**